File: Университет/cp_trainings/СНМ (DSU)/ITMO CF/step1/C/C.py
class DSU(object):
    def __init__(self, n: int) -> None:
        self.p = [i for i in range(n + 1)]
        self.size = [1] * (n + 1)
        self.e = [0] * (n + 1)
        return
    
    def __get(self, a: int) -> tuple[int, int]:
        if a == self.p[a]:
            return (a, 0)
        
        res = self.__get(self.p[a])
        self.p[a] = res[0]
        self.e[a] += res[1]
        return (self.p[a], self.e[a])
    
    def union(self, a: int, b: int) -> None:
        a = self.__get(a)[0]
        b = self.__get(b)[0]
        if a == b:
            return
        if self.size[a] < self.size[b]:
            a, b = b, a
        self.p[b] = a
        self.size[a] += self.size[b]
        self.e[b] -= self.e[a]
        return
    
    def add_exp(self, a: int, val: int) -> None:
        a = self.__get(a)[0]
        self.e[a] += val
        return
    
    def get_exp(self, a: int) -> int:
        p = self.__get(a)[0]
        if a == p:
            return self.e[a]
        return self.e[a] + self.e[p]

def main() -> None:
    n, m = map(int, input().split())
    dsu = DSU(n)
    for _ in range(m):
        inputs = input().split()
        if inputs[0] == 'join':
            x, y = int(inputs[1]), int(inputs[2])
            dsu.union(x, y)
        elif inputs[0] == 'add':
            x, v = int(inputs[1]), int(inputs[2])
            dsu.add_exp(x, v)
        else: # inputs[0] == 'get'
            x = int(inputs[1])
            print(dsu.get_exp(x))
    return

if __name__ == '__main__':
    main()

File: Университет/cp_trainings/СНМ (DSU)/ITMO CF/step1/C/C.cpp
// author: Danila "akshin_" Axyonov

#include <iostream>
#include <string>
#include <vector>
#include <algorithm>
using namespace std;
typedef long long int lli;

#define plli                pair<lli, lli>
#define mp(_first, _second) make_pair(_first, _second)
#define pb(_elem)           push_back(_elem)

#define ind                 first
#define exp                 second




class DSU {
private:
    vector<lli> p,
                size,
                e;

    plli get(lli a) {
        if (a == this->p[a])
            return mp(a, 0LL);
        plli res = this->get(this->p[a]);
        this->p[a] = res.ind;
        this->e[a] += res.exp;
        return mp(this->p[a], this->e[a]);
    }

public:
    DSU(lli n) {
        p.resize(n + 1LL);
        generate(p.begin(), p.end(), []() { static lli i = 0LL; return i++; });
        size.assign(n + 1LL, 1LL);
        e.assign(n + 1LL, 0LL);
        return;
    }

    void unite(lli a, lli b) {
        a = this->get(a).ind;
        b = this->get(b).ind;
        if (a == b)
            return;
        if (this->size[a] < this->size[b])
            swap(a, b);
        this->p[b] = a;
        this->size[a] += this->size[b];
        this->e[b] -= this->e[a];
        return;
    }

    void add_exp(lli a, lli val) {
        a = this->get(a).ind;
        this->e[a] += val;
        return;
    }

    lli get_exp(lli a) {
        lli prim = this->get(a).ind;
        if (a == prim)
            return this->e[a];
        return this->e[a] + this->e[prim];
    }
};




void solve();

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    
    solve();
    
    return 0;
}

void solve() {
    lli n,
        q;
    cin >> n >> q;
    DSU dsu(n);
    while (q--) {
        string cmd;
        cin >> cmd;
        if (cmd == "join") {
            lli x,
                y;
            cin >> x >> y;
            dsu.unite(x, y);
        } else if (cmd == "add") {
            lli x,
                v;
            cin >> x >> v;
            dsu.add_exp(x, v);
        } else { // cmd == "get"
            lli x;
            cin >> x;
            cout << dsu.get_exp(x) << '\n';
        }
    }    
    return;
}
